### PR TITLE
Metadata improvements: duration, AVI & WMV

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This adds two new features to the metadata extractor:

1) Extract duration from videos. We might want to display the whole duration of the dive in the profile (@atdotde has some interesting ideas in this regard). Moreover, in the future I want to be able to add photos/videos of a whole dive-trip at once. The file should be associated to the dive it is closest to. This is only possible for videos if we know start *and* end-time.

2) Preliminary support for AVI. Extraction of duration seemed trivial, but unfortunately the creation date does not seem to be standardized. I implemented the two versions that I found on the web. Ultimately, I fear we will have to see what our users need. Perhaps nobody uses AVI anyway. I'm not aware of a camera that outputs such a thing..!?

### Changes made:

1) Extract duration from videos.
2) Preliminary support for AVIs.
### Release note:

Perhaps?
### Documentation change:

Perhaps together with general info on video support?

### CODE REVIEW

**Comment by neolit123** [comment]
> Preliminary support for AVI. Extraction of duration seemed trivial, but unfortunately the creation date does not seem to be standardized. I implemented the two versions that I found on the web. Ultimately, I fear we will have to see what our users need. Perhaps nobody uses AVI anyway. I'm not aware of a camera that outputs such a thing..!?

would that work for other containers different from AVI?

**Comment by bstoeger** [comment]
> would that work for other containers different from AVI?

Sure - in my tests it works nicely for MP4/Quicktime derived formats. I would guess that makes up > 90% of all video files nowaday?

Indeed, the first commit *only* implements duration extraction for these formats. I thought that was obvious, so I didn't mention it.

**Comment by bstoeger** [comment]
Even though this is complex code - I think this can go into 4.8.1. The non-trivial part only affects people uploading AVIs, which we didn't even support before. I threw a some random files at it and nothing bad happened (knock wood).

File: core/metadata.cpp [hunk 1]
```diff
@@ -199,6 +222,187 @@ static bool parseMP4(QFile &f, metadata *metadata)
...
+	}
+
+	// I've also seen "Weekday Mon  Day hh:mm:ss yyyy"(!)
+	QStringList items = s.split(' ', QString::SkipEmptyParts);
+	if (items.size() < 4)
+		return false;
+
+	// Skip weekday if any is given
+	for (const QString &day: weekdays) {
+		if (items[0].startsWith(day, Qt::CaseInsensitive)) {
```
Comment (dirkhh) [inline core/metadata.cpp:255]: so this is always encoded in English abbreviation format?
That's crazy

**Comment by dirkhh** [comment]
I agree that this isn't really high risk in the sense of creating a regression. It could cause all kinds of fun issues for users playing with it, but hey...
Should this get a CHANGELOG entry that mentions "experimental support for..."?
Eventually this also needs to be mentioned in the user manual - but that doesn't need to happen for 4.8.1
I'll give others a few more hours to look and comment, but I'm willing to merge this.

File: core/metadata.cpp (hunk 1 above)
Comment (bstoeger) [reply core/metadata.cpp:255]: I think there is no standardized way to record the creation date. As far as I understood, the creating software is basically free to write in any human-readable string it wants.

I implemented the two ways I found on the web. I'd really appreciate any real-world AVI examples. But I'm not sure if any modern video equipment produces such a thing.

Comment (dirkhh) [reply core/metadata.cpp:255]: OK, makes sense. Let's go with what we have then.
I'd still appreciate the CHANGELOG entry :-)

**Comment by bstoeger** [comment]
Heh.. Already found a bug (64 bit values > 0xFFFFFFFF were not read correctly).

**Comment by bstoeger** [comment]
Well - threw in a WMV parser for good measure. By far the easiest so far. Weirdly I didn't find a .wmv file with a valid date online..?

**Comment by bstoeger** [comment]
I tested this with the following procedure
```
#include <QDir>
static void test_metadata(const QDir &dir)
{
	for (const QString &file: dir.entryList(QDir::Files)) {
		QString path = dir.filePath(file);
		printf("%s is ", qPrintable(path));
		metadata md;
		switch(get_metadata(qPrintable(path), &md)) {
		case MEDIATYPE_PICTURE:
			printf("a picture with timestamp: %ld.\n", md.timestamp);
			break;
		case MEDIATYPE_VIDEO:
			printf("a video with timestamp: %ld and duration: %d.\n", md.timestamp, md.duration.seconds);
			break;
		case MEDIATYPE_IO_ERROR:
			printf("not readable.\n");
			break;
		default:
			printf("not supported.\n");
			break;
		}
	}
	for (const QString &dirname: dir.entryList(QDir::NoDotAndDotDot | QDir::Dirs))
		test_metadata(dir.filePath(dirname));
}

extern "C" void test_metadata()
{
	test_metadata(QDir("/home/bs"));
}
```
on all files in my home directory - and believe me, this includes a lot of random garbage. It was reasonably fast and did not crash. So I think apart from erroneous extraction of video-timestamps this code should be rather safe.

**Comment by dirkhh** [comment]
merged

**Pull request merged**